cms/src/main/java/uk/nhs/digital/externalstorage/resource/ResourceUploadPlugin.java

```diff
@@ -0,0 +1,114 @@
+package uk.nhs.digital.externalstorage.resource;
+
+import static org.hippoecm.repository.api.HippoNodeType.HIPPO_TEXT;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.wicket.markup.html.form.upload.FileUpload;
+import org.hippoecm.frontend.behaviors.EventStoppingBehavior;
+import org.hippoecm.frontend.model.JcrNodeModel;
+import org.hippoecm.frontend.plugin.IPluginContext;
+import org.hippoecm.frontend.plugin.config.IPluginConfig;
+import org.hippoecm.frontend.plugins.jquery.upload.FileUploadViolationException;
+import org.hippoecm.frontend.plugins.jquery.upload.single.FileUploadPanel;
+import org.hippoecm.frontend.plugins.yui.upload.validation.DefaultUploadValidationService;
+import org.hippoecm.frontend.plugins.yui.upload.validation.FileUploadValidationService;
+import org.hippoecm.frontend.service.IEditor;
+import org.hippoecm.frontend.service.render.RenderPlugin;
+import org.onehippo.cms7.services.HippoServiceRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import uk.nhs.digital.externalstorage.ExternalStorageConstants;
+import uk.nhs.digital.externalstorage.s3.S3Connector;
+import uk.nhs.digital.externalstorage.s3.S3ObjectDescriber;
+
+import java.io.IOException;
+import java.util.Calendar;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+
+
+/**
+ * Plugin for uploading resources into the JCR repository.
+ * This plugin can be configured with specific types, so not all file types are allowed to be uploaded.
+ */
+public class ResourceUploadPlugin extends RenderPlugin {
+
+    static final Logger log = LoggerFactory.getLogger(ResourceUploadPlugin.class);
+    public static final String DEFAULT_ASSET_VALIDATION_SERVICE_ID = "service.gallery.asset.validation";
+
+    private final IEditor.Mode mode;
+
+    public ResourceUploadPlugin(IPluginContext context, IPluginConfig config) {
+        super(context, config);
+        mode = IEditor.Mode.fromString(config.getString("mode"), IEditor.Mode.EDIT);
+        add(createFileUploadPanel());
+        add(new EventStoppingBehavior("onclick"));
+    }
+
+    private FileUploadPanel createFileUploadPanel() {
+        final FileUploadPanel panel = new FileUploadPanel("fileUpload", getPluginConfig(), getValidationService()) {
+            @Override
+            public void onFileUpload(final FileUpload fileUpload) throws FileUploadViolationException {
+                handleUpload(fileUpload);
+            }
+        };
+        panel.setVisible(mode == IEditor.Mode.EDIT);
+        return panel;
+    }
+
+    private FileUploadValidationService getValidationService() {
+        return DefaultUploadValidationService.getValidationService(getPluginContext(), getPluginConfig(),
+            DEFAULT_ASSET_VALIDATION_SERVICE_ID);
```

Comment:
Assume this works for things that aren't gallery/assets?

Reply:
yes yes, this is the copy&paste from the OOTB code.